model/task_lifecycle.go

```diff
@@ -1184,25 +1192,50 @@ func UpdateDisplayTask(t *task.Task) error {
 	return nil
 }
 
+func checkResetSingleHostTaskGroup(t *task.Task, caller string) error {
+	if !t.IsPartOfSingleHostTaskGroup() {
+		return nil
+	}
+	tasks, err := task.FindTaskGroupFromBuild(t.BuildId, t.TaskGroup)
+	if err != nil {
+		return errors.Wrapf(err, "can't get task group for task '%s'", t.Id)
+	}
+	shouldReset := false
+	for _, tgTask := range tasks {
+		if tgTask.ResetWhenFinished {
+			shouldReset = true
+		}
+		if tgTask.Blocked() { // should restart if tasks are blocked
+			break
+		}
+		if !tgTask.IsFinished() && tgTask.Activated { // task is not finished and not blocked so don't restart
```

Comment:
why can't we restart when a member task in in progress? (display tasks need it to keep their execution numbers in sync)

Reply:
I can explain this better in person I think if this doesn't help:
> Why can't we abort and restart? 

We _could_ I guess, but this is a design choice: we want all tasks that can run to run first.
> Why do execution numbers need to be kept in sync?

Does it help to think about this as a whole? The purpose of this ticket is to always restart the task group _as a whole_, so if a member task is in progress we can't restart that member task, so we could mark it to restart later but there are even more edge cases with that. So we either want to restart all tasks or no tasks (until the task group is all finished).

>if a task is blocked why is it then okay? I had been thinking it could be blocked on a task outside the task group. it could be blocked on a task outside the task group.

_usually_ when a task in a single host task group is blocked, it's blocked because a previous task in a task group has failed, so it's never going to run.

I guess in the case where task 3 is blocked on a task outside the group, but no tasks in the task group have failed, but we asked to restart task 2, we have no way of knowing if task 3 will ever be unblocked, but we know any task after task 3 in the task group can't run either since task 3 is also blocked, so we may as well restart so we can run task 1 and 2 (and task 3 will still be blocked, and thus the tasks after task 3).